File: backend/app/routers/dev.py
"""Dev Door endpoints (admin-only) — docs/EVAL-DRIVEN-DEV.md.

The Prompt Inspector's dry-run: plan a chat message and return the full trace (prompt,
raw output, parsed plan, token/cost meta) WITHOUT executing anything on the server. Every
route is guarded by ``require_admin`` — never reachable by a customer, even one with a
valid token. The server is resolved through the admin's own access (team_service), so the
dev tool cannot reach a server the admin isn't entitled to.
"""
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies.auth import require_admin
from app.models.user import User
from app.services import admin_service, dev_service, team_service

router = APIRouter(prefix="/api/dev", tags=["dev"])


class DryRunRequest(BaseModel):
    server_id: uuid.UUID
    message: str = Field(min_length=1, max_length=4000)


@router.post("/dry-run")
async def dry_run(
    body: DryRunRequest,
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Plan a message as Ally would — return the full trace, never execute a command."""
    access = await team_service.get_access(db, admin, str(body.server_id))
    if access is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Server not found"
        )
    return await dev_service.dry_run(access.server, body.message, acting_user=admin)


# ── Eval runner + case capture (Phase 3 — the flywheel) ───────────────────────


class CaptureCaseRequest(BaseModel):
    category: str = Field(min_length=1, max_length=40)
    input: str = Field(min_length=1, max_length=2000)
    expected: str = Field(min_length=1, max_length=120)
    os: str = "linux"
    note: str | None = Field(default=None, max_length=1000)


def _case_out(row) -> dict:
    return {
        "id": str(row.id),
        "category": row.category,
        "input": row.input,
        "expected": row.expected,
        "os": row.os,
        "note": row.note,
        "created_at": row.created_at.isoformat() if getattr(row, "created_at", None) else None,
    }


@router.get("/evals/run")
async def evals_run(
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Run the deterministic corpus + captured cases (offline, no API, no cost)."""
    return await dev_service.run_evals(db)


@router.get("/evals/cases")
async def evals_list(
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> list[dict]:
    return [_case_out(r) for r in await dev_service.list_captured(db)]


@router.post("/evals/cases", status_code=status.HTTP_201_CREATED)
async def evals_capture(
    body: CaptureCaseRequest,
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    try:
        row = await dev_service.capture_case(
            db,
            category=body.category,
            input=body.input,
            expected=body.expected,
            os=body.os,
            note=body.note,
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))
    return _case_out(row)


@router.delete("/evals/cases/{case_id}", status_code=status.HTTP_204_NO_CONTENT)
async def evals_delete(
    case_id: uuid.UUID,
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> None:
    if not await dev_service.delete_captured(db, case_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Case not found")


# ── Observability (Phase 4) ───────────────────────────────────────────────────


@router.get("/activity")
async def activity(
    limit: int = 60,
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Recent AI calls (the ledger) + this period's cost/actions summary."""
    return await dev_service.activity(db, limit=min(max(limit, 1), 200))


# ── Operator console (SAAS-LAUNCH-PLAN §5) ───────────────────────────────────
# Support/ops, NOT billing: WHMCS owns customers, orders and revenue. These answer only
# what WHMCS cannot — servers, Ally, and what the AI really costs us. All read-only.


@router.get("/admin/overview")
async def admin_overview(
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Users, plans, active, servers, and our REAL AI cost this period."""
    return await admin_service.overview(db)


@router.get("/admin/users")
async def admin_users(
    q: str | None = None,
    limit: int = 100,
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> list[dict]:
    """Every user with plan (a read-only mirror of WHMCS) + both meters + AI cost."""
    return await admin_service.list_users(db, limit=min(max(limit, 1), 500), q=q)


@router.get("/admin/users/{user_id}")
async def admin_user_detail(
    user_id: uuid.UUID,
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """The support screen: their servers, missions and recent failures — never credentials."""
    out = await admin_service.user_detail(db, user_id)
    if out is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="No such user")
    return out


@router.get("/admin/entitlements")
async def admin_entitlements(
    limit: int = 100,
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> list[dict]:
    """"Did billing land?" — every plan change WHMCS drove."""
    return await admin_service.entitlement_log(db, limit=min(max(limit, 1), 500))


class ProviderAbRequest(BaseModel):
    # Optional OpenAI price overrides per tier, e.g. {"mid": {"in": 2.5, "out": 10}} — so
    # management can plug in OpenAI's real quote and see the true cost on our real usage.
    openai: dict[str, dict[str, float]] | None = None


@router.post("/provider-ab")
async def provider_ab(
    body: ProviderAbRequest | None = None,
    admin: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Claude vs OpenAI cost, re-priced over this period's REAL ledger token usage. No
    live calls, no OpenAI key — arithmetic over data we already have (docs/AI-METERING.md)."""
    return await dev_service.provider_ab(db, body.openai if body else None)

File: frontend/src/api/dev.ts
import { apiClient } from "./client"

/**
 * Dev Door API (docs/EVAL-DRIVEN-DEV.md) — admin-only. The dry-run plans a chat message
 * exactly as Ally would and returns the full Trace WITHOUT executing anything.
 */

export interface DryRunTrace {
  input: {
    message: string
    server: {
      id: string
      name: string
      os_type: string | null
      connection_type: string
    }
    ally_mode: string | null
    language: string
  }
  context: {
    skill: string | null
    skill_menu_offered: boolean
    has_live_snapshot: boolean
    has_scout: boolean
    has_server_profile: boolean
    has_memories: boolean
    other_servers: string | null
    use_skill_requested: string | null
  }
  prompt: {
    system: string
    volatile: string
  }
  output: {
    raw: string
    parsed: Record<string, unknown>
  }
  meta: {
    models: string[]
    calls: number
    input_tokens: number
    output_tokens: number
    cache_read_tokens: number
    cache_write_tokens: number
    cost_usd: number
    escalated: boolean
    retried_trimmed: boolean
  }
}

/** Plan a message as Ally would — returns the full trace, never executes a command. */
export async function dryRun(server_id: string, message: string): Promise<DryRunTrace> {
  const { data } = await apiClient.post<DryRunTrace>("/api/dev/dry-run", {
    server_id,
    message,
  })
  return data
}

// ── Eval runner + case capture (Phase 3) ──────────────────────────────────────

export const EVAL_CATEGORIES = [
  "skill-routing",
  "safety-block",
  "safety-confirm",
  "safety-allow",
  "readonly-allow",
  "readonly-deny",
] as const

export type EvalCategory = (typeof EVAL_CATEGORIES)[number]

export interface EvalRunResult {
  summary: { total: number; passed: number; ok: boolean }
  by_category: { category: string; passed: number; total: number }[]
  failures: {
    category: string
    input: string
    expected: string
    got: string
    error: string | null
    source: "corpus" | "captured"
  }[]
  captured: {
    id: string
    category: string
    input: string
    expected: string
    os: string
    note: string | null
    got: string
    passed: boolean
  }[]
}

export interface EvalCase {
  id: string
  category: string
  input: string
  expected: string
  os: string
  note: string | null
  created_at: string | null
}

export interface CaptureCaseBody {
  category: string
  input: string
  expected: string
  os?: string
  note?: string | null
}

/** Run the deterministic corpus + captured cases (offline, no AI cost). */
export async function runEvals(): Promise<EvalRunResult> {
  const { data } = await apiClient.get<EvalRunResult>("/api/dev/evals/run")
  return data
}

export async function listEvalCases(): Promise<EvalCase[]> {
  const { data } = await apiClient.get<EvalCase[]>("/api/dev/evals/cases")
  return data
}

export async function captureEvalCase(body: CaptureCaseBody): Promise<EvalCase> {
  const { data } = await apiClient.post<EvalCase>("/api/dev/evals/cases", body)
  return data
}

export async function deleteEvalCase(id: string): Promise<void> {
  await apiClient.delete(`/api/dev/evals/cases/${id}`)
}

// ── Observability (Phase 4) ───────────────────────────────────────────────────

export interface ActivityCall {
  created_at: string | null
  feature: string
  model: string
  skill: string | null
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cost_usd: number
  actions: number
  status: string
  user: string | null
  server: string | null
}

export interface ActivityData {
  period_start: string
  summary: { cost_usd: number; actions: number; calls: number }
  by_feature: { feature: string; cost_usd: number; calls: number }[]
  daily: { day: string; cost_usd: number; calls: number }[]
  recent: ActivityCall[]
}

/** Recent AI calls (the ledger) + this period's cost/actions summary. */
export async function getActivity(): Promise<ActivityData> {
  const { data } = await apiClient.get<ActivityData>("/api/dev/activity")
  return data
}

// ── Provider cost A/B (Claude vs OpenAI, re-priced over real ledger usage) ─────

export type OaTier = "top" | "mid" | "small"

export interface AbFeature {
  feature: string
  claude_usd: number
  openai_usd: number
  in: number
  out: number
  cache_read: number
  cache_write: number
  calls: number
}

export interface ProviderAb {
  period_start: string
  tiers: Record<OaTier, { label: string; in: number; out: number }>
  caveats: string[]
  totals: {
    claude_usd: number
    openai_usd: number
    in: number
    out: number
    cache_read: number
    cache_write: number
    cache_hit_pct: number
    delta_pct: number | null
  }
  by_feature: AbFeature[]
  model_tiers: Record<string, OaTier>
}

/** Re-price this period's real token usage on Claude vs OpenAI. Pass optional per-tier
 * OpenAI price overrides to plug in a real quote. No live calls — pure arithmetic. */
export async function getProviderAb(
  openai?: Partial<Record<OaTier, { in: number; out: number }>>,
): Promise<ProviderAb> {
  const { data } = await apiClient.post<ProviderAb>("/api/dev/provider-ab", { openai })
  return data
}

// ── Operator console (docs/SAAS-LAUNCH-PLAN.md §5) ───────────────────────────
// Support/ops only. WHMCS owns customers, orders and revenue; `plan` here is a
// read-only mirror of WHMCS's decision and is never edited from this surface.

export interface AdminOverview {
  period_start: string
  users_total: number
  users_by_plan: Record<string, number>
  users_new_this_period: number
  users_active_7d: number
  servers_total: number
  ai_cost_usd: number
  ai_actions: number
  ai_calls: number
  ai_errors: number
}

export interface AdminUser {
  id: string
  email: string
  name: string | null
  plan: string
  is_admin: boolean
  is_active: boolean
  is_verified: boolean
  created_at: string | null
  actions_used: number
  actions_limit: number
  servers_used: number
  servers_limit: number
  ai_cost_usd: number
}

export interface EntitlementEvent {
  created_at: string | null
  action: string
  email: string | null
  plan: string | null
  reference: string | null
  created: boolean | null
  forced: boolean | null
  ip: string | null
}

export interface AdminUserDetail extends AdminUser {
  totp_enabled: boolean
  preferred_language: string
  ally_mode: string
  servers: {
    id: string
    name: string
    host: string
    connection_type: string
    os_type: string | null
    status: string
    last_seen: string | null
  }[]
  missions: {
    id: string
    goal: string
    server_name: string | null
    status: string
    verified: boolean | null
    created_at: string | null
  }[]
  problems: {
    created_at: string | null
    status: string
    risk_level: string | null
    request: string
  }[]
  entitlements: EntitlementEvent[]
}

export async function getAdminOverview(): Promise<AdminOverview> {
  const { data } = await apiClient.get<AdminOverview>("/api/dev/admin/overview")
  return data
}

export async function getAdminUsers(q?: string): Promise<AdminUser[]> {
  const { data } = await apiClient.get<AdminUser[]>("/api/dev/admin/users", { params: { q } })
  return data
}

export async function getAdminUser(id: string): Promise<AdminUserDetail> {
  const { data } = await apiClient.get<AdminUserDetail>(`/api/dev/admin/users/${id}`)
  return data
}

/** "Did billing land?" — every plan change WHMCS drove. */
export async function getEntitlementLog(): Promise<EntitlementEvent[]> {
  const { data } = await apiClient.get<EntitlementEvent[]>("/api/dev/admin/entitlements")
  return data
}
